okhttp/src/main/java/okhttp3/internal/connection/Transmitter.kt

```diff
@@ -0,0 +1,341 @@
+import okhttp3.internal.Util.closeQuietly
+import okhttp3.internal.Util.sameConnection
+import okhttp3.internal.platform.Platform
+import okio.AsyncTimeout
+import okio.Timeout
+import java.io.IOException
+import java.io.InterruptedIOException
+import java.lang.ref.WeakReference
+import java.net.Socket
+import java.util.concurrent.TimeUnit.MILLISECONDS
+import javax.net.ssl.HostnameVerifier
+import javax.net.ssl.SSLSocketFactory
+
+/**
+ * Bridge between OkHttp's application and network layers. This class exposes high-level application
+ * layer primitives: connections, requests, responses, and streams.
+ *
+ * This class supports [asynchronous canceling][cancel]. This is intended to have the smallest
+ * blast radius possible. If an HTTP/2 stream is active, canceling will cancel that stream but not
+ * the other streams sharing its connection. But if the TLS handshake is still in progress then
+ * canceling may break the entire connection.
+ */
+class Transmitter(
+  private val client: OkHttpClient,
+  private val call: Call
+) {
+  private val connectionPool: RealConnectionPool = client.connectionPool().delegate
+  private val eventListener: EventListener = client.eventListenerFactory().create(call)
+  private val timeout = object : AsyncTimeout() {
+    override fun timedOut() {
+      cancel()
+    }
+  }.apply {
+    timeout(client.callTimeoutMillis().toLong(), MILLISECONDS)
+  }
+
+  private var callStackTrace: Any? = null
+
+  private var request: Request? = null
+  private var exchangeFinder: ExchangeFinder? = null
+
+  // Guarded by connectionPool.
+  var connection: RealConnection? = null
+  private var exchange: Exchange? = null
+  private var exchangeRequestDone = false
+  private var exchangeResponseDone = false
+  private var canceled = false
+  private var timeoutEarlyExit = false
+  private var noMoreExchanges = false
+
+  val isCanceled: Boolean
+    get() {
+      synchronized(connectionPool) {
+        return canceled
+      }
+    }
+
+  fun timeout(): Timeout = timeout
+
+  fun timeoutEnter() {
+    timeout.enter()
+  }
+
+  /**
+   * Stops applying the timeout before the call is entirely complete. This is used for WebSockets
+   * and duplex calls where the timeout only applies to the initial setup.
+   */
+  fun timeoutEarlyExit() {
+    check(!timeoutEarlyExit)
+    timeoutEarlyExit = true
+    timeout.exit()
+  }
+
+  private fun <E : IOException?> timeoutExit(cause: E): E {
+    if (timeoutEarlyExit) return cause
+    if (!timeout.exit()) return cause
+
+    val e = InterruptedIOException("timeout")
+    if (cause != null) e.initCause(cause)
+    @Suppress("UNCHECKED_CAST") // E is either IOException or IOException?
+    return e as E
+  }
+
+  fun callStart() {
+    this.callStackTrace = Platform.get().getStackTraceForCloseable("response.body().close()")
+    eventListener.callStart(call)
+  }
+
+  /**
+   * Prepare to create a stream to carry `request`. This prefers to use the existing connection if
```

Comment:
Should be `[request]` instead of \`request\`?

Reply:
nope

Reply:
Fixed to use `[square braces]`. My rule is to use them whenever they’re supported.